presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.plan;
+
+import com.facebook.presto.sql.planner.Symbol;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import javax.annotation.concurrent.Immutable;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+@Immutable
+public class ApplyNode
+        extends PlanNode
+{
+    private final PlanNode input;
+    private final PlanNode subquery;
+
+    // Mapping between input and subquery, key represents symbols in input, values in subquery
+    private final Map<Symbol, Symbol> correlation;
```

Comment:
I think your mental model is correct at least it is in line with my mental model. Maybe I didn't cast that into java code clearly enough. I think that comment above correlation map could be misleading, I will fix it.

The direction of mapping in correlation map to me does not have much importance. Would it be cleaner to you I would reverse it? In that case, key in map would represent some symbol in a subquery plan (inner plan) - an input to subquery call, and value in map would represent output symbol from input node (outer plan).

Take a look at the example of a real query:
`SELECT n.nationkey FROM nation n WHERE n.nationkey + 100 < (SELECT n.nationkey*2);`

This query could have below plan:
```
Project(nationkey) -> (nationkey)
  Filter(nationkey + 100 < expr) -> (nationkey, expr)
    Apply -> (nationkey, expr)
      (correlation) {nationkey: nationkey_1}
      (input) TableScan(nation) -> (nationkey)
      (subquery) EnforceSingleRow -> (expr)
        Project(expr -> nationkey_1 * 2) -> (expr)
```
At the above example, `input` of apply node returns `nationkey` symbol which is then used in `subquery` plan in place of `nationkey_1` symbol.

However, I have an idea of using plain list of correlation symbols instead of map. Then plan would like below:
```
Project(nationkey) -> (nationkey)
  Filter(nationkey + 100 < expr) -> (nationkey, expr)
    Apply -> (nationkey, expr)
      (correlation) [nationkey]
      (input) TableScan(nation) -> (nationkey)
      (subquery) EnforceSingleRow -> (expr)
        Project(expr -> nationkey * 2) -> (expr)
```

See that input and subquery plans are both using the same symbol `nationkey` (no need for `nationkey_1`). So Apply need to only know which symbols are correlated. I think it is easier to understand such plan. I will change that.